tests/python_package_test/test_basic.py

```diff
@@ -579,3 +579,30 @@ def test_param_aliases():
     assert all(len(i) >= 1 for i in aliases.values())
     assert all(k in v for k, v in aliases.items())
     assert lgb.basic._ConfigAliases.get('config', 'task') == {'config', 'config_file', 'task', 'task_type'}
+
+
+def _bad_gradients(labels, preds):
+    return np.random.randn(len(labels) + 1), np.random.rand(len(labels) + 1)
+
+
+def _good_gradients(labels, preds):
+    return np.random.randn(len(labels)), np.random.rand(len(labels))
+
+
+def test_custom_objective_safety():
+    X = np.random.randn(100, 5)
+    y_binary = np.random.choice([0, 1], 100)
+    classes = [0, 1, 2]
+    y_multiclass = np.random.choice(classes, 100)
+    ds_binary = lgb.Dataset(X, y_binary).construct()
+    ds_multiclass = lgb.Dataset(X, y_multiclass).construct()
+    bad_bst_binary = lgb.Booster({'objective': "none"}, ds_binary)
+    good_bst_binary = lgb.Booster({'objective': "none"}, ds_binary)
+    bad_bst_multi = lgb.Booster({'objective': "none", "num_class": len(classes)}, ds_multiclass)
+    good_bst_multi = lgb.Booster({'objective': "none", "num_class": len(classes)}, ds_multiclass)
+    good_bst_binary.update(fobj=_good_gradients)
+    with pytest.raises(ValueError, match="number of models per one iteration \(1\)"):
```

Comment:
```suggestion
    with pytest.raises(ValueError, match="number of models per one iteration \\(1\\)"):
```

Please see these linting errors seen in https://github.com/microsoft/LightGBM/runs/4613473477?check_suite_focus=true

> ./tests/python_package_test/test_basic.py:604:78: W605 invalid escape sequence '\('
./tests/python_package_test/test_basic.py:604:81: W605 invalid escape sequence '\)'
./tests/python_package_test/test_basic.py:607:79: W605 invalid escape sequence '\('
./tests/python_package_test/test_basic.py:607:95: W605 invalid escape sequence '\)'

Reply:
Please don't escape any symbols for the readability purpose. Just add a `*` symbol at the end:
```suggestion
    with pytest.raises(ValueError, match="number of models per one iteration (1) *"):
```

Reply:
The escape is necessary; `(` and `)` are characters which means something to the regular expression engine. I'll switch to using `re.escape`